### DISCUSSION TITLE
How to post request body with Mojo::UserAgent?

### DISCUSSION THREAD

**Discussion by aceproject** [mojolicious/mojo#2122] [Q&A] [answered]
I tried below code. But the request body are missing.

```
my $ua = Mojo::UserAgent->new;
my $url = "https://postman-echo.com/post";
my $tx = $ua->post(
    $url => {
        'Authorization' => 'Basic xxxxxxxxxxxxxx',
        'Content-Type' => 'application/x-www-form-urlencoded'
    } => form => {
        'grant_type' => 'authorization_code',
        'code' => $code,
        'xxx' => '777'
    }
);
```

Here is a result. I expected value at "form" variables.
```
'{
    "args": {},
    "data": "",
    "files": {},
    "form": {},
    "headers": {
        "x-forwarded-proto": "https",
        "x-forwarded-port": "443",
        "host": "[postman-echo.com](http://postman-echo.com/)",
        "x-amzn-trace-id": "Root=1-653c253d-2bee01d92fa16ed279d4b219",
        "content-type": "application/x-www-form-urlencoded",
        "authorization": "Basic xxxxxxxxxxxxxx"
    },
    "json": null,
    "url": "https://postman-echo.com/post"
}';
```

Anyone know the reason?

With regards,

**Comment by guest20** [comment]
add `print $tx->res->body` after that example and it does show form content on the response ...

**Comment by aceproject** [comment]
> add `print $tx->res->body` after that example and it does show form content on the response ...

$tx->res->body as same as $tx->res->text
```
'{
    "args": {},
    "data": "",
    "files": {},
    "form": {},
    "headers": {
        "x-forwarded-proto": "https",
        "x-forwarded-port": "443",
        "host": "[postman-echo.com](http://postman-echo.com/)",
        "x-amzn-trace-id": "Root=1-653c253d-2bee01d92fa16ed279d4b219",
        "content-type": "application/x-www-form-urlencoded",
        "authorization": "Basic xxxxxxxxxxxxxx"
    },
    "json": null,
    "url": "https://postman-echo.com/post"
}';
```

**Comment by aceproject** [comment, marked as answer]
Self solved. Sorry, the cause is the basic authentication. Working fine with I removed below lilne.
```
'Authorization' => 'Basic xxxxxxxxxxxxxx',
```

**Reply by guest20** [reply]
When I ran your code (with a `use 5.020; use Mojo::UserAgent; my $code=22222;` above, and a `say $tx->res->body` below) i got a:
```
  "form": {
    "code": "22222",
    "grant_type": "authorization_code",
    "xxx": "777"
  },
```
... in my response body, even with `'Authorization' => 'Basic xxxxxxxxxxxxxx',`.

It's pretty likely that:
- postman-echo rate limited you because you exceeded their rate limit out of frustration, or 
- you accidentally changed something else in the code example while you were pulling it out of your script (for example, the code you provided refers to `$code`, but you don't ever set that).
